cpp_linter/common_fs/file_filter.py

```diff
@@ -0,0 +1,216 @@
+import configparser
+import os
+from pathlib import Path, PurePath
+from typing import List, Optional, Set
+from . import FileObj
+from ..loggers import logger
+
+
+class FileFilter:
+    """A reusable mechanism for parsing and validating file filters.
+
+    :param extensions: A list of file extensions in which to focus.
+    :param ignore_value: The user input specified via :std:option:`--ignore`
+        CLI argument.
+    :param not_ignored: A list of files or paths that will be explicitly not ignored.
+    :param tool_specific_name: A clang tool name for which the file filter is
+        specifically applied. This only gets used in debug statements.
+    """
+
+    def __init__(
+        self,
+        ignore_value: str = "",
+        extensions: Optional[List[str]] = None,
+        not_ignored: Optional[List[str]] = None,
+        tool_specific_name: Optional[str] = None,
+    ) -> None:
+        #: A set of file extensions that are considered C/C++ sources.
+        self.extensions: Set[str] = set(extensions or [])
+        #: A set of ignore patterns.
+        self.ignored: Set[str] = set()
+        #: A set of not-ignore patterns.
+        self.not_ignored: Set[str] = set(not_ignored or [])
+        self._tool_name = tool_specific_name or ""
+        self._parse_ignore_option(paths=ignore_value)
+
+    def parse_submodules(self, path: str = ".gitmodules"):
+        """Automatically detect submodules from the given relative ``path``.
+        This will add each submodule to the `ignored` list unless already specified as
+        `not_ignored`."""
+        git_modules = Path(path)
+        if git_modules.exists():
+            git_modules_parent = git_modules.parent
+            submodules = configparser.ConfigParser()
+            submodules.read(git_modules.resolve().as_posix())
+            for module in submodules.sections():
+                sub_mod_path = git_modules_parent / submodules[module]["path"]
+                if not self.is_file_in_list(ignored=False, file_name=sub_mod_path):
+                    sub_mod_posix = sub_mod_path.as_posix()
+                    logger.info(
+                        "Appending submodule to ignored paths: %s", sub_mod_posix
+                    )
+                    self.ignored.add(sub_mod_posix)
+
+    def _parse_ignore_option(self, paths: str):
+        """Parse a given string of paths (separated by a ``|``) into ``ignored`` and
+        ``not_ignored`` lists of strings.
+
+        :param paths: This argument conforms to the input value of :doc:`:doc:`CLI <cli_args>` <cli_args>` arg
+            :std:option:`--ignore`.
+
+        Results are added accordingly to the `ignored` and `not_ignored` attributes.
+        """
+        if paths:
+            for path in paths.split("|"):
```

Comment:
I have to make sure that `|` is not a part of the Unix glob style syntax. If it is, then we'll have to change the pattern here (`|`) to something that isn't part of the glob style syntax (maybe like `;|`).

Reply:
- https://en.wikipedia.org/wiki/Glob_(programming)
- https://man7.org/linux/man-pages/man7/glob.7.html

Both don't mention `|` as part of the Unix style glob syntax. I think we're good to go here?